File: ts/src/multiDevice.ts
import type { DeviceEntry } from "./AppKeys"
import type { InvitePurpose } from "./Invite"

export type AppKeysSnapshotDecision =
  | "advanced"
  | "stale"
  | "merged_equal_timestamp"

type MergeableAppKeys<T> = {
  merge: (other: T) => T
}

export interface AppKeysSnapshotUpdate<T> {
  decision: AppKeysSnapshotDecision
  appKeys: T
  createdAt: number
}

export interface ApplyAppKeysSnapshotOptions<T extends MergeableAppKeys<T>> {
  currentAppKeys?: T | null
  currentCreatedAt?: number
  incomingAppKeys: T
  incomingCreatedAt: number
}

export function applyAppKeysSnapshot<T extends MergeableAppKeys<T>>(
  options: ApplyAppKeysSnapshotOptions<T>
): AppKeysSnapshotUpdate<T> {
  const {
    currentAppKeys,
    currentCreatedAt = 0,
    incomingAppKeys,
    incomingCreatedAt,
  } = options

  if (!currentAppKeys || incomingCreatedAt > currentCreatedAt) {
    return {
      decision: "advanced",
      appKeys: incomingAppKeys,
      createdAt: incomingCreatedAt,
    }
  }

  if (incomingCreatedAt < currentCreatedAt) {
    return {
      decision: "stale",
      appKeys: currentAppKeys,
      createdAt: currentCreatedAt,
    }
  }

  return {
    decision: "merged_equal_timestamp",
    appKeys: currentAppKeys.merge(incomingAppKeys),
    createdAt: currentCreatedAt,
  }
}

export interface DeviceRegistrationState {
  isCurrentDeviceRegistered: boolean
  hasKnownRegisteredDevices: boolean
  noPreviousDevicesFound: boolean
  requiresDeviceRegistration: boolean
  canSendPrivateMessages: boolean
}

export interface EvaluateDeviceRegistrationStateOptions {
  currentDevicePubkey?: string | null
  registeredDevices: Array<Pick<DeviceEntry, "identityPubkey">>
  hasLocalAppKeys?: boolean
  appKeysManagerReady?: boolean
  sessionManagerReady?: boolean
}

export function evaluateDeviceRegistrationState(
  options: EvaluateDeviceRegistrationStateOptions
): DeviceRegistrationState {
  const {
    currentDevicePubkey,
    registeredDevices,
    hasLocalAppKeys = false,
    appKeysManagerReady = false,
    sessionManagerReady = false,
  } = options

  const normalizedCurrent = currentDevicePubkey?.trim().toLowerCase() ?? null
  const isCurrentDeviceRegistered =
    normalizedCurrent !== null &&
    registeredDevices.some(
      (device) => device.identityPubkey.trim().toLowerCase() === normalizedCurrent
    )

  const hasKnownRegisteredDevices = registeredDevices.length > 0

  return {
    isCurrentDeviceRegistered,
    hasKnownRegisteredDevices,
    noPreviousDevicesFound: hasKnownRegisteredDevices === false,
    requiresDeviceRegistration:
      normalizedCurrent !== null && isCurrentDeviceRegistered === false,
    canSendPrivateMessages:
      appKeysManagerReady &&
      sessionManagerReady &&
      (hasLocalAppKeys || isCurrentDeviceRegistered || hasKnownRegisteredDevices),
  }
}

export function shouldRequireRelayRegistrationConfirmation(
  options: EvaluateDeviceRegistrationStateOptions
): boolean {
  const state = evaluateDeviceRegistrationState(options)
  return state.requiresDeviceRegistration && state.hasKnownRegisteredDevices
}

type SessionStateLike = {
  theirCurrentNostrPublicKey?: string
  theirNextNostrPublicKey?: string
}

type SessionLike = {
  state?: SessionStateLike | null
}

type SessionDeviceLike = {
  activeSession?: SessionLike | null
  inactiveSessions?: Array<SessionLike | null>
}

type SessionAppKeysLike = {
  getAllDevices?: () => Array<{
    identityPubkey?: string | null
  }>
}

export type SessionUserRecordLike = {
  devices?: Map<string, SessionDeviceLike>
  appKeys?: SessionAppKeysLike | null
}

export type SessionUserRecordsLike = Map<string, SessionUserRecordLike>

export function isOwnDevicePubkey(
  pubkey: string,
  ownerPubkey: string,
  identityPubkey: string | null,
  devices: DeviceEntry[]
): boolean {
  if (!pubkey) return false
  if (pubkey === ownerPubkey) return true
  if (identityPubkey && pubkey === identityPubkey) return true
  return devices.some((device) => device.identityPubkey === pubkey)
}

export function isOwnDeviceEvent(
  eventPubkey: string,
  sessionPubkey: string,
  ownerPubkey: string,
  identityPubkey: string | null,
  devices: DeviceEntry[]
): boolean {
  return (
    isOwnDevicePubkey(eventPubkey, ownerPubkey, identityPubkey, devices) ||
    isOwnDevicePubkey(sessionPubkey, ownerPubkey, identityPubkey, devices)
  )
}

export function hasExistingSessionWithRecipient(
  userRecords: SessionUserRecordsLike | null | undefined,
  recipientPubkey: string
): boolean {
  if (!userRecords || !recipientPubkey) return false

  for (const [recordPubkey, userRecord] of userRecords.entries()) {
    const devices = userRecord?.devices
    if (!devices) continue

    for (const device of devices.values()) {
      const sessions = [device.activeSession, ...(device.inactiveSessions ?? [])]
      for (const session of sessions) {
        if (!session) continue
        const state = session.state
        if (!state) continue

        if (
          recordPubkey === recipientPubkey ||
          state.theirCurrentNostrPublicKey === recipientPubkey ||
          state.theirNextNostrPublicKey === recipientPubkey
        ) {
          return true
        }
      }
    }
  }

  return false
}

export function resolveSessionPubkeyToOwner(
  userRecords: SessionUserRecordsLike | null | undefined,
  pubkey: string
): string {
  if (!userRecords || !pubkey) return pubkey

  for (const [recordPubkey, userRecord] of userRecords.entries()) {
    if (recordPubkey === pubkey) {
      return recordPubkey
    }

    const devices = userRecord?.devices
    if (devices?.has(pubkey)) {
      return recordPubkey
    }

    const appKeyDevices = userRecord?.appKeys?.getAllDevices?.() ?? []
    if (appKeyDevices.some((device) => device.identityPubkey === pubkey)) {
      return recordPubkey
    }

    if (!devices) continue

    for (const device of devices.values()) {
      const sessions = [device.activeSession, ...(device.inactiveSessions ?? [])]
      for (const session of sessions) {
        if (!session) continue
        const state = session.state
        if (!state) continue

        if (
          state.theirCurrentNostrPublicKey === pubkey ||
          state.theirNextNostrPublicKey === pubkey
        ) {
          return recordPubkey
        }
      }
    }
  }

  return pubkey
}

export type RumorLike = {
  pubkey: string
  tags?: string[][]
}

function firstTagValue(tags: string[][] | undefined, name: string): string | undefined {
  return tags?.find((tag) => tag[0] === name)?.[1]
}

export function resolveRumorPeerPubkey(options: {
  ownerPubkey: string
  rumor: RumorLike
  senderPubkey?: string | null
}): string | undefined {
  const normalizedOwner = options.ownerPubkey.trim().toLowerCase()
  const normalizedRumorPubkey = options.rumor.pubkey.trim().toLowerCase()
  const normalizedSenderPubkey = options.senderPubkey?.trim().toLowerCase()

  if (
    normalizedRumorPubkey === normalizedOwner ||
    normalizedSenderPubkey === normalizedOwner
  ) {
    return firstTagValue(options.rumor.tags, "p")
  }

  return options.rumor.pubkey
}

export function resolveConversationCandidatePubkeys(options: {
  ownerPubkey: string
  rumor: RumorLike
  senderPubkey: string
}): string[] {
  const owner = options.ownerPubkey.trim().toLowerCase()
  const sender = options.senderPubkey.trim().toLowerCase()
  const rumorAuthor = options.rumor.pubkey.trim().toLowerCase()
  const pTagPubkey = firstTagValue(options.rumor.tags, "p")?.trim().toLowerCase()

  const isSelfTargetedRumor =
    (rumorAuthor === owner || sender === owner) &&
    (pTagPubkey == null || pTagPubkey.length === 0 || pTagPubkey === owner)

  const candidates: string[] = []
  const addCandidate = (candidate?: string | null) => {
    const normalized = candidate?.trim().toLowerCase()
    if (!normalized || candidates.includes(normalized)) return
    candidates.push(normalized)
  }

  if (isSelfTargetedRumor) {
    if (rumorAuthor !== owner) addCandidate(rumorAuthor)
    if (sender !== owner) addCandidate(sender)
    addCandidate(owner)
    return candidates
  }

  addCandidate(
    resolveRumorPeerPubkey({
      ownerPubkey: owner,
      rumor: options.rumor,
      senderPubkey: sender,
    })
  )
  addCandidate(sender)
  return candidates
}

export interface InviteOwnerRoutingResolution {
  ownerPublicKey: string
  claimedOwnerPublicKey: string
  verifiedWithAppKeys: boolean
  usedLinkBootstrapException: boolean
  fellBackToDeviceIdentity: boolean
}

export function resolveInviteOwnerRouting(options: {
  devicePubkey: string
  claimedOwnerPublicKey?: string | null
  invitePurpose?: InvitePurpose
  currentOwnerPublicKey: string
  appKeys?: {
    getAllDevices: () => Array<Pick<DeviceEntry, "identityPubkey">>
  } | null
}): InviteOwnerRoutingResolution {
  const devicePubkey = options.devicePubkey.trim().toLowerCase()
  const claimedOwnerPublicKey =
    options.claimedOwnerPublicKey?.trim().toLowerCase() || devicePubkey

  if (claimedOwnerPublicKey === devicePubkey) {
    return {
      ownerPublicKey: devicePubkey,
      claimedOwnerPublicKey,
      verifiedWithAppKeys: claimedOwnerPublicKey === devicePubkey,
      usedLinkBootstrapException: false,
      fellBackToDeviceIdentity: false,
    }
  }

  const verifiedWithAppKeys =
    options.appKeys?.getAllDevices().some((device) => device.identityPubkey === devicePubkey) ??
    false
  if (verifiedWithAppKeys) {
    return {
      ownerPublicKey: claimedOwnerPublicKey,
      claimedOwnerPublicKey,
      verifiedWithAppKeys: true,
      usedLinkBootstrapException: false,
      fellBackToDeviceIdentity: false,
    }
  }

  const usedLinkBootstrapException =
    options.invitePurpose === "link" &&
    claimedOwnerPublicKey === options.currentOwnerPublicKey.trim().toLowerCase()
  if (usedLinkBootstrapException) {
    return {
      ownerPublicKey: claimedOwnerPublicKey,
      claimedOwnerPublicKey,
      verifiedWithAppKeys: false,
      usedLinkBootstrapException: true,
      fellBackToDeviceIdentity: false,
    }
  }

  return {
    ownerPublicKey: devicePubkey,
    claimedOwnerPublicKey,
    verifiedWithAppKeys: false,
    usedLinkBootstrapException: false,
    fellBackToDeviceIdentity: true,
  }
}

File: rust/crates/nostr-double-ratchet/src/multi_device.rs
use crate::{is_app_keys_event, AppKeys};
use nostr::{Event, PublicKey};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum AppKeysSnapshotDecision {
    Advanced,
    Stale,
    MergedEqualTimestamp,
}

#[derive(Debug, Clone)]
pub struct AppKeysSnapshot {
    pub decision: AppKeysSnapshotDecision,
    pub app_keys: AppKeys,
    pub created_at: u64,
}

pub fn apply_app_keys_snapshot(
    current_app_keys: Option<&AppKeys>,
    current_created_at: u64,
    incoming_app_keys: &AppKeys,
    incoming_created_at: u64,
) -> AppKeysSnapshot {
    if current_app_keys.is_none() || incoming_created_at > current_created_at {
        return AppKeysSnapshot {
            decision: AppKeysSnapshotDecision::Advanced,
            app_keys: incoming_app_keys.clone(),
            created_at: incoming_created_at,
        };
    }

    let current_app_keys = current_app_keys.expect("checked above");
    if incoming_created_at < current_created_at {
        return AppKeysSnapshot {
            decision: AppKeysSnapshotDecision::Stale,
            app_keys: current_app_keys.clone(),
            created_at: current_created_at,
        };
    }

    AppKeysSnapshot {
        decision: AppKeysSnapshotDecision::MergedEqualTimestamp,
        app_keys: current_app_keys.merge(incoming_app_keys),
        created_at: current_created_at,
    }
}

pub fn select_latest_app_keys_from_events<'a, I>(events: I) -> Option<AppKeysSnapshot>
where
    I: IntoIterator<Item = &'a Event>,
{
    let mut latest: Option<AppKeysSnapshot> = None;

    for event in events {
        if !is_app_keys_event(event) {
            continue;
        }

        let Ok(app_keys) = AppKeys::from_event(event) else {
            continue;
        };

        latest = Some(match latest.as_ref() {
            Some(current) => apply_app_keys_snapshot(
                Some(&current.app_keys),
                current.created_at,
                &app_keys,
                event.created_at.as_secs(),
            ),
            None => AppKeysSnapshot {
                decision: AppKeysSnapshotDecision::Advanced,
                app_keys,
                created_at: event.created_at.as_secs(),
            },
        });
    }

    latest
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct DeviceRegistrationState {
    pub is_current_device_registered: bool,
    pub has_known_registered_devices: bool,
    pub no_previous_devices_found: bool,
    pub requires_device_registration: bool,
    pub can_send_private_messages: bool,
}

pub fn evaluate_device_registration_state(
    current_device_pubkey_hex: Option<&str>,
    registered_device_pubkeys: &[String],
    has_local_app_keys: bool,
    app_keys_manager_ready: bool,
    session_manager_ready: bool,
) -> DeviceRegistrationState {
    let normalized_current = current_device_pubkey_hex
        .map(str::trim)
        .map(str::to_lowercase)
        .filter(|value| !value.is_empty());

    let is_current_device_registered = normalized_current.as_ref().is_some_and(|current| {
        registered_device_pubkeys
            .iter()
            .any(|device| device.trim().eq_ignore_ascii_case(current))
    });
    let has_known_registered_devices = !registered_device_pubkeys.is_empty();

    DeviceRegistrationState {
        is_current_device_registered,
        has_known_registered_devices,
        no_previous_devices_found: !has_known_registered_devices,
        requires_device_registration: normalized_current.is_some() && !is_current_device_registered,
        can_send_private_messages: app_keys_manager_ready
            && session_manager_ready
            && (has_local_app_keys || is_current_device_registered || has_known_registered_devices),
    }
}

pub fn should_require_relay_registration_confirmation(
    current_device_pubkey_hex: Option<&str>,
    registered_device_pubkeys: &[String],
    has_local_app_keys: bool,
    app_keys_manager_ready: bool,
    session_manager_ready: bool,
) -> bool {
    let state = evaluate_device_registration_state(
        current_device_pubkey_hex,
        registered_device_pubkeys,
        has_local_app_keys,
        app_keys_manager_ready,
        session_manager_ready,
    );
    state.requires_device_registration && state.has_known_registered_devices
}

fn normalize_pubkey_hex(value: &str) -> Option<String> {
    let normalized = value.trim().to_lowercase();
    if normalized.is_empty() {
        None
    } else {
        Some(normalized)
    }
}

fn first_tag_value(tags: &[Vec<String>], name: &str) -> Option<String> {
    tags.iter()
        .find(|tag| tag.first().is_some_and(|tag_name| tag_name == name) && tag.len() >= 2)
        .and_then(|tag| normalize_pubkey_hex(&tag[1]))
}

pub fn resolve_rumor_peer_pubkey(
    owner_pubkey_hex: &str,
    rumor_pubkey_hex: &str,
    rumor_tags: &[Vec<String>],
    sender_pubkey_hex: Option<&str>,
) -> Option<String> {
    let normalized_owner = normalize_pubkey_hex(owner_pubkey_hex)?;
    let normalized_rumor_pubkey = normalize_pubkey_hex(rumor_pubkey_hex)?;
    let normalized_sender_pubkey = sender_pubkey_hex.and_then(normalize_pubkey_hex);

    if normalized_rumor_pubkey == normalized_owner
        || normalized_sender_pubkey.as_deref() == Some(normalized_owner.as_str())
    {
        return first_tag_value(rumor_tags, "p");
    }

    Some(normalized_rumor_pubkey)
}

pub fn resolve_conversation_candidate_pubkeys(
    owner_pubkey_hex: &str,
    rumor_pubkey_hex: &str,
    rumor_tags: &[Vec<String>],
    sender_pubkey_hex: &str,
) -> Vec<String> {
    let Some(owner) = normalize_pubkey_hex(owner_pubkey_hex) else {
        return Vec::new();
    };
    let Some(sender) = normalize_pubkey_hex(sender_pubkey_hex) else {
        return Vec::new();
    };
    let Some(rumor_author) = normalize_pubkey_hex(rumor_pubkey_hex) else {
        return Vec::new();
    };
    let p_tag_pubkey = first_tag_value(rumor_tags, "p");

    let is_self_targeted_rumor = (rumor_author == owner || sender == owner)
        && match p_tag_pubkey.as_deref() {
            Some(p_tag) => p_tag.is_empty() || p_tag == owner,
            None => true,
        };

    let mut candidates = Vec::new();
    let mut add_candidate = |candidate: Option<String>| {
        let Some(candidate) = candidate else {
            return;
        };
        if !candidates.contains(&candidate) {
            candidates.push(candidate);
        }
    };

    if is_self_targeted_rumor {
        if rumor_author != owner {
            add_candidate(Some(rumor_author));
        }
        if sender != owner {
            add_candidate(Some(sender));
        }
        add_candidate(Some(owner));
        return candidates;
    }

    add_candidate(resolve_rumor_peer_pubkey(
        owner.as_str(),
        rumor_pubkey_hex,
        rumor_tags,
        Some(sender.as_str()),
    ));
    add_candidate(Some(sender));
    candidates
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct InviteOwnerRoutingResolution {
    pub owner_pubkey: PublicKey,
    pub claimed_owner_pubkey: PublicKey,
    pub verified_with_app_keys: bool,
    pub used_link_bootstrap_exception: bool,
    pub fell_back_to_device_identity: bool,
}

pub fn resolve_invite_owner_routing(
    device_pubkey: PublicKey,
    claimed_owner_pubkey: PublicKey,
    invite_purpose: Option<&str>,
    current_owner_pubkey: PublicKey,
    app_keys: Option<&AppKeys>,
) -> InviteOwnerRoutingResolution {
    if claimed_owner_pubkey == device_pubkey {
        return InviteOwnerRoutingResolution {
            owner_pubkey: device_pubkey,
            claimed_owner_pubkey,
            verified_with_app_keys: true,
            used_link_bootstrap_exception: false,
            fell_back_to_device_identity: false,
        };
    }

    let verified_with_app_keys = app_keys
        .and_then(|keys| keys.get_device(&device_pubkey))
        .is_some();
    if verified_with_app_keys {
        return InviteOwnerRoutingResolution {
            owner_pubkey: claimed_owner_pubkey,
            claimed_owner_pubkey,
            verified_with_app_keys: true,
            used_link_bootstrap_exception: false,
            fell_back_to_device_identity: false,
        };
    }

    let used_link_bootstrap_exception =
        invite_purpose == Some("link") && claimed_owner_pubkey == current_owner_pubkey;
    if used_link_bootstrap_exception {
        return InviteOwnerRoutingResolution {
            owner_pubkey: claimed_owner_pubkey,
            claimed_owner_pubkey,
            verified_with_app_keys: false,
            used_link_bootstrap_exception: true,
            fell_back_to_device_identity: false,
        };
    }

    InviteOwnerRoutingResolution {
        owner_pubkey: device_pubkey,
        claimed_owner_pubkey,
        verified_with_app_keys: false,
        used_link_bootstrap_exception: false,
        fell_back_to_device_identity: true,
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::DeviceEntry;
    use nostr::{EventBuilder, Keys, Kind, Timestamp};

    #[test]
    fn merges_same_second_app_keys_snapshots_monotonically() {
        let device1 = Keys::generate().public_key();
        let device2 = Keys::generate().public_key();
        let current = AppKeys::new(vec![DeviceEntry::new(device1, 100)]);
        let incoming = AppKeys::new(vec![
            DeviceEntry::new(device1, 100),
            DeviceEntry::new(device2, 100),
        ]);

        let applied = apply_app_keys_snapshot(Some(&current), 100, &incoming, 100);

        assert_eq!(
            applied.decision,
            AppKeysSnapshotDecision::MergedEqualTimestamp
        );
        assert!(applied.app_keys.get_device(&device1).is_some());
        assert!(applied.app_keys.get_device(&device2).is_some());
    }

    #[test]
    fn selects_newest_app_keys_event_by_created_at() {
        let owner_keys = Keys::generate();
        let device1 = Keys::generate().public_key();
        let device2 = Keys::generate().public_key();
        let sign_app_keys = |app_keys: AppKeys, created_at: u64| {
            let unsigned = app_keys.get_event(owner_keys.public_key());
            EventBuilder::new(Kind::from(crate::APP_KEYS_EVENT_KIND as u16), "")
                .tags(unsigned.tags.clone())
                .custom_created_at(Timestamp::from(created_at))
                .build(owner_keys.public_key())
                .sign_with_keys(&owner_keys)
                .unwrap()
        };

        let older_event = sign_app_keys(AppKeys::new(vec![DeviceEntry::new(device1, 100)]), 100);
        let newer_event = sign_app_keys(
            AppKeys::new(vec![
                DeviceEntry::new(device1, 100),
                DeviceEntry::new(device2, 101),
            ]),
            101,
        );

        let selected = select_latest_app_keys_from_events([&older_event, &newer_event])
            .expect("expected latest AppKeys snapshot");

        assert_eq!(selected.decision, AppKeysSnapshotDecision::Advanced);
        assert_eq!(selected.created_at, 101);
        assert!(selected.app_keys.get_device(&device1).is_some());
        assert!(selected.app_keys.get_device(&device2).is_some());
    }

    #[test]
    fn falls_back_to_device_identity_for_unverified_chat_invites() {
        let device = Keys::generate().public_key();
        let owner = Keys::generate().public_key();
        let current_owner = Keys::generate().public_key();

        let resolved =
            resolve_invite_owner_routing(device, owner, Some("chat"), current_owner, None);

        assert_eq!(resolved.owner_pubkey, device);
        assert!(resolved.fell_back_to_device_identity);
    }

    #[test]
    fn keeps_owner_side_link_bootstrap_before_appkeys_registration() {
        let device = Keys::generate().public_key();
        let owner = Keys::generate().public_key();

        let resolved = resolve_invite_owner_routing(device, owner, Some("link"), owner, None);

        assert_eq!(resolved.owner_pubkey, owner);
        assert!(resolved.used_link_bootstrap_exception);
    }

    #[test]
    fn resolves_self_targeted_conversation_candidates_with_linked_device_first() {
        let candidates = resolve_conversation_candidate_pubkeys(
            "owner",
            "linked-device",
            &[vec!["p".into(), "owner".into()]],
            "owner",
        );

        assert_eq!(
            candidates,
            vec!["linked-device".to_string(), "owner".to_string()]
        );
    }

    #[test]
    fn evaluates_device_registration_state() {
        let state = evaluate_device_registration_state(
            Some("device-2"),
            &[String::from("device-1")],
            false,
            true,
            true,
        );

        assert!(!state.is_current_device_registered);
        assert!(state.has_known_registered_devices);
        assert!(!state.no_previous_devices_found);
        assert!(state.requires_device_registration);
        assert!(state.can_send_private_messages);
    }

    #[test]
    fn skips_relay_confirmation_for_first_device_bootstrap() {
        assert!(!should_require_relay_registration_confirmation(
            Some("device-1"),
            &[],
            false,
            true,
            true,
        ));
    }

    #[test]
    fn requires_relay_confirmation_for_new_device_on_existing_owner() {
        assert!(should_require_relay_registration_confirmation(
            Some("device-2"),
            &[String::from("device-1")],
            false,
            true,
            true,
        ));
    }
}
